x/capability/keeper/keeper.go

```diff
@@ -0,0 +1,244 @@
+package keeper
+
+import (
+	"fmt"
+
+	"github.com/tendermint/tendermint/libs/log"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/capability/types"
+)
+
+type (
+	// Keeper defines the capability module's keeper. It is responsible for provisioning,
+	// tracking, and authenticating capabilities at runtime. During application
+	// initialization, the keeper can be hooked up to modules through unique function
+	// references so that it can identify the calling module when later invoked.
+	//
+	// When the initial state is loaded from disk, the keeper allows the ability to
+	// create new capability keys for all previously allocated capability identifiers
+	// (allocated during execution of past transactions and assigned to particular modes),
+	// and keep them in a memory-only store while the chain is running.
+	//
+	// The keeper allows the ability to create scoped sub-keepers which are tied to
+	// a single specific module.
+	Keeper struct {
+		cdc           *codec.Codec
+		storeKey      sdk.StoreKey
+		memKey        sdk.StoreKey
+		scopedModules map[string]struct{}
+		sealed        bool
+	}
+
+	// ScopedKeeper defines a scoped sub-keeper which is tied to a single specific
+	// module provisioned by the capability keeper. Scoped keepers must be created
+	// at application initialization and passed to modules, which can then use them
+	// to claim capabilities they receive and retrieve capabilities which they own
+	// by name, in addition to creating new capabilities & authenticating capabilities
+	// passed by other modules.
+	ScopedKeeper struct {
+		cdc      *codec.Codec
+		storeKey sdk.StoreKey
+		memKey   sdk.StoreKey
+		module   string
+	}
+)
+
+func NewKeeper(cdc *codec.Codec, storeKey, memKey sdk.StoreKey) Keeper {
+	return Keeper{
+		cdc:           cdc,
+		storeKey:      storeKey,
+		memKey:        memKey,
```

Comment:
I'd `panic` here if the `memKey.Type()` doesn't match `sdk. MemoryStoreKey`

Reply:
We could, but the app has to call `InitializeAndSeal` which does the check without us having to change the interface of `StoreKey`.